src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,820 @@
+			// to prevent wasting space on them.
+			// TODO(rartoul):This is an uncommon scenario and this operation might
+			// be expensive to perform each time so consider removing this if it
+			// impacts performance too much.
+			fieldNum := field.GetNumber()
+			if !fieldsContains(fieldNum, schemaFields) {
+				if err := m.TryClearFieldByNumber(int(fieldNum)); err != nil {
+					return fmt.Errorf(
+						"error: %v clearing field that does not exist in schema: %d", err, fieldNum)
+				}
+			}
+		}
+
+		for _, field := range schemaFields {
+			var (
+				fieldNum    = field.GetNumber()
+				fieldNumInt = int(fieldNum)
+				prevVal     = enc.lastEncoded.GetFieldByNumber(fieldNumInt)
+				curVal      = m.GetFieldByNumber(fieldNumInt)
+			)
+
+			if fieldsEqual(curVal, prevVal) {
+				// Clear fields that haven't changed.
+				if err := m.TryClearFieldByNumber(fieldNumInt); err != nil {
+					return fmt.Errorf("error: %v clearing field: %d", err, fieldNumInt)
+				}
+			} else {
+				isDefaultValue, err := isDefaultValue(field, curVal)
+				if err != nil {
+					return fmt.Errorf(
+						"error: %v, checking if %v is default value for field %s",
+						err, curVal, field.String())
+				}
+				if isDefaultValue {
+					fieldsChangedToDefault = append(fieldsChangedToDefault, fieldNum)
+				}
+
+				changedFields = append(changedFields, fieldNum)
+				if err := enc.lastEncoded.TrySetFieldByNumber(fieldNumInt, curVal); err != nil {
+					return fmt.Errorf(
+						"error: %v setting field %d with value %v on lastEncoded",
+						err, fieldNumInt, curVal)
+				}
+			}
+		}
+	}
+
+	if len(changedFields) == 0 && enc.lastEncoded != nil {
+		// Only want to skip encoding if nothing has changed AND we've already
+		// encoded the first message.
+		enc.stream.WriteBit(opCodeNoChange)
+		return nil
+	}
+
+	// TODO(rartoul): Need to add a MarshalInto to the ProtoReflect library to save
+	// allocations: https://github.com/m3db/m3/issues/1471
+	marshaled, err := m.Marshal()
+	if err != nil {
+		return fmt.Errorf("%s error trying to marshal protobuf: %v", encErrPrefix, err)
+	}
+
+	// Control bit indicating that proto values have changed.
+	enc.stream.WriteBit(opCodeChange)
+	if len(fieldsChangedToDefault) > 0 {
+		// Control bit indicating that some fields have been set to default values
+		// and that a bitset will follow specifying which fields have changed.
+		enc.stream.WriteBit(opCodeFieldsSetToDefaultProtoMarshal)
+		enc.encodeBitset(fieldsChangedToDefault)
+	} else {
+		// Control bit indicating that none of the changed fields have been set to
+		// their default values so we can do a clean merge on read.
+		enc.stream.WriteBit(opCodeNoFieldsSetToDefaultProtoMarshal)
+	}
+	enc.encodeVarInt(uint64(len(marshaled)))
+	enc.stream.WriteBytes(marshaled)
+
+	if enc.lastEncoded == nil {
+		// Set lastEncoded to m so that subsequent encodings only need to encode fields
+		// that have changed.
+		enc.lastEncoded = dynamic.NewMessage(enc.schema)
```

Comment:
There's no `Reset()` like method?  Sucks that this has to be allocated again and again.

Can this be fixed upstream?